Make Permalinks editable

pento opened the pull request:
## Description

It's time to make permalinks editable.

Fixes #1285.

## Screenshots (jpeg or gifs if applicable):

<img width="692" alt="viewing permalink" src="https://user-images.githubusercontent.com/352291/32658726-8141ae60-c66f-11e7-87fb-ea57541ae492.png">
<img width="664" alt="editing permalink" src="https://user-images.githubusercontent.com/352291/32658728-81785b90-c66f-11e7-8932-9b62f86666eb.png">

pento committed:
Add a sample permalink to post responses

pento committed:
Show the sample permalink, when available.

pento committed:
Linting errors

pento committed:
Add the sample permalink to all post types

pento committed:
Open the slug edit form when clicking on the slug

pento committed:
Fix some linting errors

afercia commented:
There are a few accessibility concerns with this UI, starting from the placement in the source: it's difficult to tab to it, the only way is to tab backwards. When tabbing forwards, the whole component is rendered when the tab sequence has already "skipped" it so assistive technologies users won't have a clue something appeared on the screen.

The button to edit should be a separate control: currently the button is announced based on the slug value, this doesn't give a proper feedback about the button action. The input field is unlabeled. There's no "cancel". In the current interface, the permalink is a fully functional link and it's clickable. Instead, now it's just a span with text styled as a link.

However, the first thing to check is probably the input field is actually not editable:
> Warning: Failed form propType: You provided a `value` prop to a form field without an `onChange` handler. This will render a read-only field. If the field should be mutable use `defaultValue`. Otherwise, set either `onChange` or `readOnly`.

About the UI and interaction, I'd recommend to try to replicate the current one in the Classic editor. There are good reason why each action has a dedicated control. For reference:

![screen shot 2017-11-10 at 16 06 10](https://user-images.githubusercontent.com/1682452/32665939-29d903dc-c636-11e7-912b-bb1529c4b3f8.png)

<img width="689" alt="screen shot 2017-11-10 at 16 07 05" src="https://user-images.githubusercontent.com/1682452/32665940-29f2922a-c636-11e7-9c4f-5761f9c6f49e.png">

![screen shot 2017-11-10 at 16 13 05](https://user-images.githubusercontent.com/1682452/32665941-2a0fcfe8-c636-11e7-816a-397f231d5955.png)

afercia commented:
P.S. the demo page crashes when clicking on some blocks, I guess because there's no slug there?

pento commented:
Yeah, I quickly hacked this into the existing UI to see how it feels - I'm not particularly enamoured by it. It feels super hidden, you can really only discover it by accident.

@karmatosed, what are your thoughts on a slug editing UI?

aduth commented on editor/post-permalink/index.js:
```diff
@@ -42,21 +46,68 @@ class PostPermalink extends Component {
+			viewLink = link;
+		if ( samplePermalink ) {
+			permalink = samplePermalink[ 0 ].replace( '%postname%', samplePermalink[ 1 ] );
+			viewLink += '&preview=true';
```
Can we know for certain that the `viewLink` already includes other query parameters? Maybe we ought to use `addQueryArgs` from the `@wordpress/url` package?

https://github.com/WordPress/packages/blob/master/packages/url/README.md

karmatosed commented:
I have a suggestion (a little bit out there maybe), but I would like to get feedback from @jasmussen on this. If we can, I would like us to use the format we have right now. Here is my suggestion for that without clicking in to make active:

![1](https://user-images.githubusercontent.com/253067/32805646-f48b1f0e-c981-11e7-8207-0333be033b86.png)

The second is when you click in to edit:

![2](https://user-images.githubusercontent.com/253067/32805732-3b0af422-c982-11e7-8efd-25885ea0e62a.png)

aduth commented on editor/post-permalink/index.js:
```diff
@@ -42,21 +46,68 @@ class PostPermalink extends Component {
+					{ ! editingSlug &&
+						<Button
+							className="editor-post-permalink__slug"
+							onClick={ this.onEditPermalink }
```
We shouldn't offer the option to edit permalink if `%postname%` doesn't exist in the permalink structure.

aduth commented on editor/post-permalink/index.js:
```diff
@@ -42,21 +46,68 @@ class PostPermalink extends Component {
+			viewLink += '&preview=true';
+		}
+
+		const prefix = permalink.replace( /[^/]+\/?$/, '' ),
```
Can you explain these regular expressions, preferably in a code comment?

aduth commented on editor/post-permalink/index.js:
```diff
@@ -42,21 +46,68 @@ class PostPermalink extends Component {
 	}
 
+	onEditPermalink( event ) {
+		event.preventDefault();
```
Why do we need to prevent default here?

aduth commented on editor/post-permalink/index.js:
```diff
@@ -42,21 +46,68 @@ class PostPermalink extends Component {
+	}
+
+	onCancelEditPermalink( event ) {
+		event.preventDefault();
```
Why do we need to prevent default here?

jasmussen commented:
@karmatosed I like your screenshot, I think it's an enhancement.

However I do agree with both you, Pento and Andrea that this UI (showing the permalink when clicking the title field) is a failed experiment. I can say that because it was my idea ¯\_(ツ)_/¯

There are some alternative designs in the ticket that I think we should explore (including [the link icon](https://github.com/WordPress/gutenberg/issues/1285#issuecomment-309997645). 

So yes, we absolutely need to make a better UI for it. However the question is whether we need to do it as part of **this** PR. Perhaps we can do it in a two-parter — first part being to get this PR in, so it's at least editable, and then we make sure to reopen #1285 (or open a new separate ticket) to improve the UI.

afercia commented:
Just to note that in the classic editor, the visible permalink is a fully functional link and it's clickable. Personally, I use it a lot of times to view the post in a new tab, because in the current UI is the most prominent "preview" link for me.
In the new editor this will change, because the permalink UI will be displayed only when needed, but I'd consider to keep the link clickable. The last part shouldn't be a button or input field by default. It should transform in an input field when editing and there should be a separate "Edit" button.

karmatosed commented:
Thanks for inputs. I am going to take this on and look at what a next version would be design wise. I'll update here once have that.

karmatosed commented:
I did a little digging into this and went back to the idea of what we have currently:

![permalinks-existing](https://user-images.githubusercontent.com/253067/33241555-f0278f92-d2be-11e7-97f8-a36144b0c6a9.png)

Right now copying is by selecting and copying. We also have a click to view. I thought about what we have in Gutenberg:

![permalinks-gutenberg](https://user-images.githubusercontent.com/253067/33241558-feea11da-d2be-11e7-8e8c-5a37b5eace16.png)

This is nice because of the simplification. However it loses the ability to edit easily and copying is a step. What about combining?

![permalinks-suggestion](https://user-images.githubusercontent.com/253067/33241563-1acbfec2-d2bf-11e7-844f-be7829ed80a1.png)

The flow to this would be:
- Click to change permalinks: this is a useful feature for users and I think worth putting back in, at least until we can find a better place for it.
- Editing is a one step with an 'ok' being closing and cancelling. This works as both get the same end result. Saving would happen only if a change had happened. Over having a cancel button.
- Clicking the permalink shows it just like in the original version.

The benefit of this is we keep the minimal UI but we get back some of the features expected.

@jasmussen what do you think of this as a suggestion?

jasmussen commented:
> @jasmussen what do you think of this as a suggestion?

I dig it, and any improvements here are welcome. I don't know that the original issue is solved, though, around discoverability.

Perhaps we should revisit the design that had a floating icon symbol next to the title.

karmatosed commented:
@jasmussen true, it does solve the editable issue though. I do think this icon has a potential:

![permalink1](https://user-images.githubusercontent.com/253067/33282093-6c20d2f8-d39f-11e7-8195-b930ec810fcc.png)

jasmussen commented:
I think the stock dashicons link icon could work, yes. Should it be vertically centered in the title box?

karmatosed commented:
@pento are you up for iterating on this?

amandablum commented:
how can I help? (other than to test, which I'm happy to do)
![Image of Scary Mervyn's Lady](https://tenor.com/view/mervyns-open-hurry-gif-5538300)

pento committed:
Add a link button beside the post title, to open the permalink UI.

pento committed:
Use an edit button to edit the permalink.

pento committed:
Use addQueryArgs() for adding the preview argument.

pento committed:
Fix a PHP warning

karmatosed commented:
This seems ready to me to get merged with what we have. One potential improvement would be for the link icon to be a copy of the url. I actually found myself wanting to do that. Right now you can't highlight and copy from within Gutenberg the url easily, it clicks to show and you can't copy.

![2018-01-10 at 11 04](https://user-images.githubusercontent.com/253067/34769614-3f577406-f5f6-11e7-9c7c-75165032d24e.png)

afercia commented:
Apart from accessibility concerns to address, there are still some functional and design issues.
- the slug field is still not editable (JS warning missing onChange handler)
- I have no idea how the permalink gets saved 🙂 
- when a notice is displayed, for example "Post updated") the whole UI is hidden behind the notice, only a tiny part of the button is barely visible:

![screen shot 2018-01-10 at 19 43 43](https://user-images.githubusercontent.com/1682452/34789534-eff56a00-f63e-11e7-8cee-c35e4418587a.png)

Re: the accessibility issues, I'd like to remind UI controls should be grouped in a logical order so I'd recommend to not make the left icon a "Copy" button (it would also miss the confirmation message that appears in place of the original button text). I'd rather place two buttons on the right. Other changes will be necessary to address a11y but I'd say to make it functional first and then improve the a11y part.

schlessera commented:
It looks like the permalink is editable even if permalinks are not being used:
<img width="683" alt="image 2018-01-11 at 4 03 59 pm" src="https://user-images.githubusercontent.com/83631/34832079-27563af6-f6e9-11e7-953e-2e726572b594.png">
This should obviously not be possible, I guess the Edit button should then be greyed out or hidden.

afercia commented:
In the Classic editor, when the permalink setting is set to "Plain", the UI is not editable and shows just a button (actually a link) to the Permalink Settings page:

<img width="507" alt="screen shot 2018-01-11 at 16 52 59" src="https://user-images.githubusercontent.com/1682452/34833566-07d27454-f6f0-11e7-8eb9-5700abf04ca5.png">

afercia commented:
Current "logic": https://core.trac.wordpress.org/browser/trunk/src/wp-admin/includes/post.php?rev=42343#L1367

Looking a that, there are at least a couple more things to address:
- it should check also for `'%pagename%'` and not just for `'%postname%'`
- when it's a page and it's the front page, the permalink is not editable and doesn't show the link to the permalinks settings

schlessera committed:
Remove unused variables.

schlessera committed:
Fix indentation linting error.

schlessera committed:
Make the slug editable.

schlessera committed:
Use post's slug property directly.

schlessera commented:
If pushed some changes to the PR to make the editing of the permalink work all the way through persistence.

Right now, the validation is solely done on the server end. The slug is pushed through the REST API, and the server then sanitizes the slug. It returns a modified post object, that contains the (potentially adapted) slug.

The Permalink component waits for that response and then adapts its state to accept whatever the server provided. So, as an example, if you already have a post called `test`, and you rename another post to `test` as well, it will change to `test-2` a few moments after saving the post.

An obvious improvement would be to add basic sanitization at the client level, but proper sanitization would require roundtrips to the server. I think that this implementation is good enough for a first basic implementation, and further UI/UX iterations should be done before delving too deep into the sanitization rabbit hole.

pento committed:
Fix a bunch of eslint errors.

pento committed:
Remove some unused code.

pento committed:
Add a couple more jsdoc blocks

schlessera committed:
Reintroduce "Copy" button next to permalink

The removal seems to have been unintended.

schlessera committed:
Adapt the buttons to share the same styling

Note: the `<Button>` component does not add the `button` class by default, so I added it manually. See https://github.com/WordPress/gutenberg/issues/4638

schlessera committed:
Add docblock to onCopy handler

pento committed:
Fix up some docblocks

pento committed:
Use the actual slug for published posts.

pento commented:
It's still a bit buggy, but this PR has been going on way too long. I'm going to merge it, then we can iterate on it a bit more in subsequent PRs.

karmatosed commented:
There are a number of issues with this that are blockers to releasing.

Firstly, if you make an edit, then save, it doesn't take until the second time you edit. Aside from that there are a number of visual issues that diverge form the designs:

- The input area is out of line.
- There are 2 buttons : copy and edit beside each other the same styling.

gziolo commented:
Reverted with #4675 to proceed with 2.1 release.

afercia commented:
I guess the branch should be restored?

schlessera commented:
I can continue work on this based on the recent feedback. Just a couple of questions:

* Will the branch be restored or shall I create a new PR?
* Is there a reference somewhere of what components are meant to be used, and when or how? It is difficult to find out what can be reused to base new code on.

aduth commented:
I can restore the branch, though it's my understanding that it's not possible to reopen a merged pull request within GitHub. I'm unsure whether a prompt will become available to create a new pull request from the same branch.

>Is there a reference somewhere of what components are meant to be used, and when or how? It is difficult to find out what can be reused to base new code on.

If the components are within the same top-level folder, all components can be accessed via relative paths (`import from '../../'`). For other top-level folders, it depends on whether it's exported (usually from its `index.js`), but generally I'd say most components are fair to use. That said, there's a loose hierarchy of decreasing specificity from `edit-post` (post editor) > `editor` (general block editor) > `blocks` (blocks API, library), and these components should ideally not be consuming upwards in the hierarchy (i.e. the generalized block editor `editor` should not be using components from the post editor `edit-post`).

schlessera commented:
Continued in #5414 .
